src/ProtoBuf.FSharp/CodeGen.fs

```diff
@@ -1,13 +1,44 @@
 module internal ProtoBuf.FSharp.CodeGen
 
 open System
-open FSharp.Reflection
 open System.Collections.Concurrent
+open System.Collections.Generic
+open System.Threading
 open System.Reflection
 open System.Reflection.Emit
+open FSharp.Reflection
 open MethodHelpers
 
-let inline emitZeroValueOntoEvaluationStack (gen: ILGenerator) (getterType: MethodType) =
+
+type private TypeBuilder with
+    member tb.DefineOpExplicit(src : Type, dst : Type) =
+        let attr = MethodAttributes.Public ||| MethodAttributes.HideBySig ||| MethodAttributes.SpecialName ||| MethodAttributes.Static
+        tb.DefineMethod("op_Explicit", attr, dst, [| src |])
+
+    member tb.SetProtoContractAttribute(skipConstructor : bool) =
+        let t = typeof<ProtoBuf.ProtoContractAttribute>
+        CustomAttributeBuilder(
+            t.GetConstructor [||], [||]
+            , [| t.GetProperty "ImplicitFields" ; t.GetProperty "SkipConstructor" |]
+            , [| box ProtoBuf.ImplicitFields.AllFields ; box skipConstructor |]
+        ) |> tb.SetCustomAttribute
+
+    member tb.DefineFieldForProtobuf(fi : PropertyInfo) =
+        tb.DefineField(fi.Name, fi.PropertyType, FieldAttributes.Public) // Do something with name and attributes?
+
+type private FieldBuilder with
+    member fb.SetProtoMemberAttribute(tag : int) =
+        let t = typeof<ProtoBuf.ProtoMemberAttribute>
+        CustomAttributeBuilder(t.GetConstructor [| typeof<int> |], [| box tag |]) |> fb.SetCustomAttribute
+
+
+let private emitDefaultValueViaCell (gen : ILGenerator) (tp : Type) =
+    let cell = gen.DeclareLocal(tp)
+    gen.Emit(OpCodes.Ldloca_S, cell)
+    gen.Emit(OpCodes.Initobj, tp)
+    gen.Emit(OpCodes.Ldloc, cell)
```

Comment:
While I understand the need for some Reflection.Emit I'm getting a little worried that the amount being incorporated into this PR will make the library somewhat unmaintainable/unapproachable to newcomers. Its also somewhat admittedly why its taken me quite a while to do start reviewing this PR. Reflection.Emit programming at this level is not something people do everyday.

At the very least we should comment each line describing how it does what it does making it quicker to read or a story of what this part is doing. Most maintainers approaching this won't be Reflection.Emit specialists. I would be happy with a comment on each line for now. As an example:

```/// This function generates a local variable, pushes the address into a variable, and then creates a new object there, loading it back into the variable. Intended to be used to assign default values into constructors of generated surrogates.```

Its starting to get to the point where, maybe not in this PR, I am starting to think FSharp.Compiler.Service may be something to look into as time permits or the TypeProviders SDK with its type generator DSL. Or some sort of DSL to generate types on top.

Something like:
```
let surrogateTypeBuilder = createType "SurrogateTypeForClassX"
let endType = 
   fields 
   |> Seq.fold surrogateType (fun st fi -> surrogateType |> addField fi)
```